### ISSUE TITLE
Is `ApproxNoOp` "dead code" and therefore should be removed?

### ISSUE THREAD

**Issue by andrewvaughanj** [cvc5/cvc5#6968]
In discussing #6958, @4tXJ7f pointed out that that the `--use-approx` code only is triggered if we have `options::useApprox()` and `ApproximateSimplex::enabled()`:

- https://github.com/cvc5/cvc5/blob/e4e19cd62b3eebed2de5b9b627509df0ffec23e1/src/theory/arith/theory_arith_private.cpp#L1884-L1885
- https://github.com/cvc5/cvc5/blob/e4e19cd62b3eebed2de5b9b627509df0ffec23e1/src/theory/arith/theory_arith_private.cpp#L2969

If this is the case, then it means that [`ApproximateSimplex::mkApproximateSimplexSolver`](https://github.com/cvc5/cvc5/blob/bd33d20609999f6f847aeb63a42350aeb3041406/src/theory/arith/approx_simplex.cpp#L538) is only called if we have GLPK.

If `ApproximateSimplex::mkApproximateSimplexSolver` is only called only GLPK-enabled builds, then this means that `return new ApproxNoOp(vars, l, s);` is never invoked:

- https://github.com/cvc5/cvc5/blob/bd33d20609999f6f847aeb63a42350aeb3041406/src/theory/arith/approx_simplex.cpp#L539-L543

and therefore the whole of `ApproxNoOp` is never actually used in `cvc5`.

Does this mean that this is a candidate for removal? If so, I'm happy to put this in a PR and clean-up  `ApproximateSimplex::mkApproximateSimplexSolver` -- I guess we need to decide what it should do if someone mistakenly calls it in non-GLPK builds (assert? return NULL?).

**Comment by 4tXJ7f** [comment]
From what I can tell, you are right. Before each call to `ApproximateSimplex::mkApproximateSimplexSolver`, we assert that `ApproximateSimplex::enabled()`, which is only true if cvc5 is compiled with GLPK. I think that we can remove `ApproxNoOp`.

**Comment by andrewvaughanj** [comment]
The question is though, what should `mkApproximateSimplexSolver` do if we're in a non-GLPK build? Return `nullptr`? `Assert(false)`? The call to `mkApproximateSimplexSolver` isn't `#ifdef'`d out, so it has to do _something_ if we're not building with GLPK.

**Comment by 4tXJ7f** [comment]
Ah right, I think that something like `Unimplemented() << "Approximate simplex solver requires GLPK"` would do the trick.

**Comment by andrewvaughanj** [comment]
Is it also worth “flattening” the inheritance on `ApproximateSimplex` and only having one the GLPK-focussed class (`ApproxGLPK`)?

I think this is another instance where `ApproximateSimplex` only exists to be extended by `ApproxGLPK` and has no other uses otherwise.

**Comment by andrewvaughanj** [comment]
Yeah, I think maybe it should be retained as an interface class (to ensure “cleanliness” of the users) but all the behaviour moved into the GLPK class.

I’ll take a look and make a PR!

**Comment by aytey** [comment]
Closed in https://github.com/cvc5/cvc5/commit/f1f106ed8900a8b952797d93485149122a9206d3
